pkg/git/factory/gitfactory.go

```diff
@@ -2,43 +2,76 @@ package gitfactory
 
 import (
 	"context"
+	"fmt"
+	"path/filepath"
 
 	"github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	"github.com/aws/eks-anywhere/pkg/filewriter"
 	"github.com/aws/eks-anywhere/pkg/git"
+	"github.com/aws/eks-anywhere/pkg/git/gitclient"
 	"github.com/aws/eks-anywhere/pkg/git/gogithub"
 	"github.com/aws/eks-anywhere/pkg/git/providers/github"
 )
 
-type gitProviderFactory struct {
-	GitClient github.GitClient
-}
+type gitProviderFactory struct{}
 
-type Options struct {
-	GithubGitClient github.GitClient
+func New() *gitProviderFactory {
+	return &gitProviderFactory{}
 }
 
-func New(opts Options) *gitProviderFactory {
-	return &gitProviderFactory{GitClient: opts.GithubGitClient}
+func (g *gitProviderFactory) Build(ctx context.Context, cluster *v1alpha1.Cluster, fluxConfig *v1alpha1.FluxConfig, writer filewriter.FileWriter) (git.ProviderClient, git.Client, filewriter.FileWriter, error) {
+	var providerClient git.ProviderClient
+	var gitClient git.Client
+	var repoWriter filewriter.FileWriter
+	var err error
+
+	switch {
+	case fluxConfig.Spec.Github != nil:
+		providerClient, gitClient, repoWriter, err = g.buildGithubProvider(ctx, cluster.Name, fluxConfig.Spec.Github.Owner, fluxConfig.Spec.Github.Repository, fluxConfig.Spec.Github.Personal, writer)
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("building github provider: %v", err)
+		}
+	default:
+		return nil, nil, nil, fmt.Errorf("no valid git provider in FluxConfigSpec. Spec: %v", fluxConfig)
+	}
+	return providerClient, gitClient, repoWriter, nil
 }
 
-// BuildProvider will configure and return the proper Github based on the given GitOps configuration.
-func (g *gitProviderFactory) BuildProvider(ctx context.Context, fluxConfig *v1alpha1.FluxConfigSpec) (git.Provider, error) {
+func (g *gitProviderFactory) buildGithubProvider(ctx context.Context, clusterName string, owner string, repo string, personal bool, writer filewriter.FileWriter) (git.ProviderClient, git.Client, filewriter.FileWriter, error) {
```

Comment:
what abt passing `github *GithubProviderConfig` as arg instead of `owner, repo, personal`?

Reply:
donezo 26afe51e482afff7fd14a95c0a3411544dfcb319